cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,234 @@
+
+    :param path: string representing a Windows path. The root drive must exist,
+    although the path is not required to.
+    """
+    out, _ = subp.subp([WSLPATH_CMD, "-au", path])
+    return PurePath(out.rstrip())
+
+
+def cmd_executable() -> PurePath:
+    """
+    Returns the Linux path to the Windows host's cmd.exe.
+    """
+
+    mounts = mounted_win_drives()
+    if not mounts:
+        raise IOError("Windows drives are not mounted.")
+
+    # cmd.exe path is being stable for decades.
+    candidate = "%s/Windows/System32/cmd.exe"
+    for mnt in mounts:
+        cmd = candidate % mnt
+        if not os.access(cmd, os.X_OK):
+            continue
+
+        LOG.debug("Found cmd.exe at <%s>", cmd)
+        return PurePath(cmd)
+
+    raise IOError(
+        "Couldn't find cmd.exe in any mount point: %s" % ", ".join(mounts)
+    )
+
+
+def win_user_profile_dir() -> PurePath:
+    """
+    Returns the Windows user profile directory translated as a Linux path
+    accessible inside the current WSL instance.
+    """
+    cmd = cmd_executable()
+
+    # cloud-init runs too early to rely on binfmt to execute Windows binaries.
+    # But we know that `/init` is the interpreter, so we can run it directly.
+    # See /proc/sys/fs/binfmt_misc/WSLInterop[-late]
+    # inside any WSL instance for more details.
+    home, _ = subp.subp(["/init", cmd.as_posix(), "/C", "echo %USERPROFILE%"])
+    home = home.rstrip()
+    if not home:
+        raise subp.ProcessExecutionError(
+            "No output from cmd.exe to show the user profile dir."
+        )
+
+    return win_path_2_wsl(home.rstrip())
+
+
+def machine_id():
+    """
+    Returns the local machine ID value from /etc/machine-id.
+    """
+    from cloudinit.settings import MACHINE_ID_FILE
+
+    if util.wait_for_files([MACHINE_ID_FILE], 2.0):
+        LOG.debug("%s file not found", MACHINE_ID_FILE)
+        return None
+
+    return util.load_file(MACHINE_ID_FILE, decode=True)
+
+
+def candidate_user_data_file_names(instance_name) -> List[str]:
+    """
+    Return a list of candidate file names that may contain user-data
+    in some supported format, ordered by precedence.
+    """
+    distribution_id, _, release_codename = util.get_linux_distro()
+
+    return [
+        # WSL instance specific:
+        "%s.user-data" % instance_name,
+        # release codename specific
+        "%s-%s.user-data" % (distribution_id, release_codename),
+        # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
+        "%s-all.user-data" % distribution_id,
+        # generic, valid for all WSL distros and instances.
+        "config.user-data",
+    ]
+
+
+class DataSourceWSL(sources.DataSource):
+    dsname = "WSL"
+
+    def __init__(self, sys_cfg, distro, paths):
+        sources.DataSource.__init__(self, sys_cfg, distro, paths)
+        self._network_config = sources.UNSET
+        self.dsmode = sources.DSMODE_LOCAL
+        self.distro = distro
```

Comment:
I don't think we need/want any of these attributes set in __init__for a couple reasons:
1. _network_config is only referenced in datasources that try to cache the read network config and override the DataSource.network_config property to return the cached value instead of recalculating it
2. dsmode is unused in this DataSource and was used only in other datasources to print specific information about what bode the datasource was detected in. Since this datasource is only detected in local stage per the `datasources: ....WSL, sources.DEP_FILESYSTEM` registration at the bottom, we don't really care about tracking and reporting the dsmode in any string representation of this class.
3. self.distro is set in the superclass `__init__` anyway
4. We probably don't want self.instance_id setup in __init__ because next reboot of this instance will leverage the cached datasource in /var/lib/cloud/instance/obj.pkl and will not re-run __init__ when deserializing the pickled datasource cache. So, if the instance name changed across reboot, the _get_data calls will be using the old instance name instead if the instance name has the ability to change across reboots.


I think we should drop the __init_ from DataSourceWSL and also push self.instance_name calculdation down into `_get_data()` which will be called every boot.